src/stridelayout.jl

```diff
@@ -0,0 +1,269 @@
+
+stride_rank(::Type{B}) where {T,N,I1,I2,A<:AbstractArray{T,N},B<:PermutedDimsArray{T,N,I1,I2,A}} = _stride_rank(B, stride_rank(A))
+_stride_rank(::Type{B}, ::Nothing) where {T,N,I1,I2,A<:AbstractArray{T,N},B<:PermutedDimsArray{T,N,I1,I2,A}} = nothing
+_stride_rank(::Type{B}, rank) where {T,N,I1,I2,A<:AbstractArray{T,N},B<:PermutedDimsArray{T,N,I1,I2,A}} = rank[Val{I1}()]
+function stride_rank(::Type{S}) where {N,NP,T,A<:AbstractArray{T,NP},I,S <: SubArray{T,N,A,I}}
+    _stride_rank(S, stride_rank(A))
+end
+_stride_rank(::Any, ::Any) = nothing
+@generated function _stride_rank(::Type{S}, ::StrideRank{R}) where {R,N,NP,T,A<:AbstractArray{T,NP},I,S <: SubArray{T,N,A,I}}
+    rankv = collect(R)
+    rank_new = Int[]
+    n = 0
+    for np in 1:NP
+        r = rankv[np]
+        if I.parameters[np] <: AbstractUnitRange
+            n += 1
+            push!(rank_new, r)
+        end
+    end
+    # If n != N, then an axis was indeced by something other than an integer or `AbstractUnitRange`, so we return `nothing`
+    n == N || return nothing
+    ranktup = Expr(:tuple); append!(ranktup.args, rank_new) # dynamic splats bad
+    Expr(:call, Expr(:curly, :StrideRank, ranktup))
+end
+stride_rank(x, i) = stride_rank(x)[i]
+
+struct DenseDims{D} end
+Base.@pure DenseDims(D::NTuple{<:Any,Bool}) = DenseDims{D}()
+@inline Base.getindex(::DenseDims{D}, i::Integer) where {D} = D[i]
+@inline Base.getindex(::DenseDims{D}, ::Val{I}) where {D,I} = DenseDims{permute(D, I)}()
+"""
+dense_dims(::Type{T}) -> NTuple{N,Bool}
+
+Returns a tuple of indicators for whether each axis is dense.
+An axis `i` of array `A` is dense if `stride(A, i) * size(A, i) == stride(A, j)` where `stride_rank(A)[i] + 1 == stride_rank(A)[j]`.
+"""
+dense_dims(x) = dense_dims(typeof(x))
+dense_dims(::Type) = nothing
+dense_dims(::Type{Array{T,N}}) where {T,N} = DenseDims{ntuple(_ -> true, Val{N}())}()
+dense_dims(::Type{<:Tuple}) = DenseDims{(true,)}()
+function dense_dims(::Type{<:Union{Transpose{T,A},Adjoint{T,A}}}) where {T,A<:AbstractMatrix{T}}
+    dense = dense_dims(A)
+    isnothing(dense) ? nothing : dense[Val{(2,1)}()]
+end
+function dense_dims(::Type{<:PermutedDimsArray{T,N,I1,I2,A}}) where {T,N,I1,I2,A<:AbstractArray{T,N}}
+    dense = dense_dims(A)
+    isnothing(dense) ? nothing : dense[Val{I1}()]
+end
+function dense_dims(::Type{S}) where {N,NP,T,A<:AbstractArray{T,NP},I,S <: SubArray{T,N,A,I}}
+    _dense_dims(S, dense_dims(A), stride_rank(A))
+end
+_dense_dims(::Any, ::Any) = nothing
+@generated function _dense_dims(::Type{S}, ::DenseDims{D}, ::StrideRank{R}) where {D,R,N,NP,T,A<:AbstractArray{T,NP},I,S <: SubArray{T,N,A,I}}
+    still_dense = true
+    sp = rank_to_sortperm(R)
+    densev = Vector{Bool}(undef, NP)
+    for np in 1:NP
+        spₙ = sp[np]
+        still_dense &= D[spₙ]
+        densev[spₙ] = still_dense
+        # a dim not being complete makes later dims not dense
+        still_dense && (still_dense = (I.parameters[spₙ] <: Base.Slice)::Bool)
```

Comment:
```suggestion
        still_dense &= (I.parameters[spₙ] <: Base.Slice)::Bool)
```
Should be the same and simpler?